File: backend/backup/feature-flag-service/go/internal/repository/feature_flag_repository.go
package repository

import (
	"context"

	"github.com/codebridge/feature-flag-service/internal/model"
)

// FeatureFlagRepository defines the interface for feature flag storage
type FeatureFlagRepository interface {
	// GetFlag gets a flag by key and namespace
	GetFlag(ctx context.Context, key, namespace string) (*model.FeatureFlag, error)
	
	// GetFlags gets multiple flags by keys and namespace
	GetFlags(ctx context.Context, keys []string, namespace string) ([]*model.FeatureFlag, error)
	
	// SetFlag sets a flag
	SetFlag(ctx context.Context, flag *model.FeatureFlag) error
	
	// DeleteFlag deletes a flag
	DeleteFlag(ctx context.Context, key, namespace string) error
	
	// ListFlags lists flags by namespace and optional prefix
	ListFlags(ctx context.Context, namespace, prefix string) ([]*model.FeatureFlag, error)
	
	// ListFlagsByTag lists flags by tag
	ListFlagsByTag(ctx context.Context, namespace, tagKey, tagValue string) ([]*model.FeatureFlag, error)
	
	// Subscribe subscribes to flag updates
	Subscribe(ctx context.Context, callback func(*model.FeatureFlag)) error
	
	// Unsubscribe unsubscribes from flag updates
	Unsubscribe(ctx context.Context) error
}

File: backend/backup/feature-flag-service/python/feature_flag_service/repository/feature_flag_repository.py
"""
Feature flag repository interface for the Feature Flag Service.
"""

from abc import ABC, abstractmethod
from typing import Callable, List, Optional

from feature_flag_service.model.feature_flag import FeatureFlag


class FeatureFlagRepository(ABC):
    """
    Interface for feature flag storage.
    """

    @abstractmethod
    async def get_flag(self, key: str, namespace: str) -> Optional[FeatureFlag]:
        """
        Get a flag by key and namespace.
        """
        pass

    @abstractmethod
    async def get_flags(self, keys: List[str], namespace: str) -> List[FeatureFlag]:
        """
        Get multiple flags by keys and namespace.
        """
        pass

    @abstractmethod
    async def set_flag(self, flag: FeatureFlag) -> None:
        """
        Set a flag.
        """
        pass

    @abstractmethod
    async def delete_flag(self, key: str, namespace: str) -> None:
        """
        Delete a flag.
        """
        pass

    @abstractmethod
    async def list_flags(self, namespace: str, prefix: Optional[str] = None) -> List[FeatureFlag]:
        """
        List flags by namespace and optional prefix.
        """
        pass

    @abstractmethod
    async def list_flags_by_tag(self, namespace: str, tag_key: str, tag_value: str) -> List[FeatureFlag]:
        """
        List flags by tag.
        """
        pass

    @abstractmethod
    async def subscribe(self, callback: Callable[[FeatureFlag], None]) -> None:
        """
        Subscribe to flag updates.
        """
        pass

    @abstractmethod
    async def unsubscribe(self) -> None:
        """
        Unsubscribe from flag updates.
        """
        pass

